streams/runners/storm/sampling/src/main/java/com/hortonworks/streamline/streams/sampling/service/storm/StormTopologySamplingService.java

```diff
@@ -0,0 +1,257 @@
+        Integer length = null;
+        if (!events.isEmpty()) {
+            if (events.size() > qps.count()) {
+                int i = qps.desc() ? events.size() - qps.count() : 0;
+                int j = qps.desc() ? events.size() : qps.count();
+                events = events.subList(i, j);
+            }
+            if (qps.desc()) {
+                if (qps.start() != null) {
+                    next = qps.start() + events.get(0).getStartOffset() - qps.length() + 1;
+                } else if (prevOffset != null) {
+                    next = prevOffset + events.get(0).getStartOffset() + 1;
+                }
+                if (next >= 0) {
+                    length = qps.length();
+                } else if (qps.start() > 0 && qps.start() < qps.length()){
+                    length = events.get(0).getStartOffset() + qps.start().intValue() + 1;
+                } else {
+                    length = events.get(0).getStartOffset() + 1;
+                }
+                next = Math.max(next, 0);
+                Collections.reverse(events);
+            } else {
+                SampledEvent lastEvent = events.get(events.size() - 1);
+                next = (qps.start() == null ? 0 : qps.start()) + lastEvent.getStartOffset() + lastEvent.getLength() - 1;
+                length = qps.length() == null ? Integer.valueOf(BYTES_TO_FETCH) : qps.length();
+            }
+        }
+        return buildSampledEvents(next, length, events);
+    }
+
+    private SampledEvents buildSampledEvents(Long next, Integer length, List<SampledEvent> events) {
+        return new SampledEvents() {
+            @Override
+            public Collection<SampledEvent> getEvents() {
+                return events;
+            }
+
+            @Override
+            public Long getNext() {
+                return next;
+            }
+
+            @Override
+            public Integer getLength() {
+                return length;
+            }
+        };
+    }
+
+    private void addSampledEvents(String lines, String queriedComponentId, List<SampledEvent> events) {
+        // Timestamp, Component name, Component task-id, MessageId (incase of anchoring), List of emitted values
+        int offset = 0;
+        for (String line : lines.split("\n")) {
+            line = StringEscapeUtils.unescapeHtml(line);
+            int componentStart = line.indexOf(',') + 1;
+            if (componentStart != -1) {
+                int componentEnd = line.indexOf(',', componentStart);
+                if (componentEnd != -1) {
+                    String componentName = line.substring(componentStart, componentEnd);
+                    try {
+                        if (componentName.equals(queriedComponentId)) {
+                            long time = Date.parse(line.substring(0, componentStart - 1));
+                            String eventStr = line.substring(line.indexOf(TO_STRING_PREFIX) + TO_STRING_PREFIX.length(), line.lastIndexOf("]"));
```

Comment:
Someone may think it is not necessary, but for me it's better to do defensive programming: finding TO_STRING_PREFIX and "]" before extracting eventStr. If it can't find any of them, leave log message and skip the line.

Reply:
And I just found the case above.

Reply:
Other case than splitted event, now we are relying on actual class (StreamlineEventImpl) rather than interface, so if someone has a need to use other than StreamlineEventImpl, the logic breaks. 
Unfortunately I need to have the implementation other than StreamlineEventImpl due to preserve ID while modifying event, and if we don't want to break the logic then before emitting we should convert it to StreamlineEventImpl again. Losing ID at that time may not matter since we are just emitting, but I'm nervous to add conversion logic to all the emitting places.

Reply:
Will change prefix to `StreamlineEvent` so that `IdPreservedStreamlineEvent` is also covered.